### PR TITLE
TripleLift Prebid S2S Adapter 

### PR DESCRIPTION
**Pull request by Kevin-P-Kerr**
The TripleLift prebid S2S adapter.  

request to prebid
```
{ "id": "some-request-id",
   "site": {
    "page": "https://www.123greetings.com/events/fathers_day/happy_fathers_day",
    "publisher": {
      "id": "pub-4627517680249670",
      "ext": {
        "country": "US"
      }
    } 
    },
    "imp": [
      {
        "tagid" : "makersalley_plain",
        "id": "some-impression-id",
        "banner": {
          "format": [
            {
              "w": 300,
              "h": 250
            },
            {
              "w": 300,
              "h": 600
            }
          ]
        },
        "ext": {
          "triplelift" : { "inventoryCode":"makersalley_main"} 
        }
      }
    ],
    "tmax": 500
  }
```
request from prebid to tl
```
"{
  "id": "some-request-id",
  "imp": [
    {
      "id": "some-impression-id",
      "banner": {
        "format": [
          {
            "w": 300,
            "h": 250
          },
          {
            "w": 300,
            "h": 600
          }
        ]
      },
      "tagid": "makersalley_plain",
      "ext": {
        "bidder": {
          "inventoryCode": "makersalley_main"
        }
      }
    }
  ],
  "site": {
    "page": "https://www.123greetings.com/events/fathers_day/happy_fathers_day",
    "publisher": {
      "id": "pub-4627517680249670",
      "ext": {
        "country": "US"
      }
    },
    "ext": {
      "amp": 0
    }
  },
  "device": {
    "ua": "curl/7.43.0",
    "ip": "::1"
  },
  "at": 1,
  "tmax": 500
}"
```
response from prebid
```
"{
  "id": "some-request-id",
  "seatbid": [
    {
      "bid": [
        {
          "id": "21524535241656531860",
          "impid": "some-impression-id",
          "price": 3.338,
          "adm": "<script>document.createElement('IMG').src=\"https://tlx.3lift.net/s2s/notify?px=1&ts=1560870964&aid=21524535241656531860&ec=260_9_2763332&n=GqsBaHR0cDovL2V4Y2hhbmdlLXRlc3Rlci50cmlwbGVsaWZ0Lm5ldDo4MDc3L25vdGlmeT9pZD0yMTUyNDUzNTI0MTY1NjUzMTg2MCZwcmljZT0ke0FVQ1RJT05fUFJJQ0V9JmFiaWQ9MTU2MzA5ODUzODIwODY3NjYwNyZhaW1waWQ9MTU2MzA5ODUzODIwODY3NjYwNy0xLW51bGwmdHlwZT1CaWQyMlNlYXRz8gKeAQgAEhQyMTUyNDUzNTI0MTY1NjUzMTg2MBgAIAEohAIwCUABSABQAGABaABwmKMFkAEAmAEAqAH%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F8BsAGPKLgBAMABihrIAY8o4AGKGvABAPgB0imAAogbiAKKGpECAAAAAAAA8D%2BZAmZmZmZmZtY%2FoQIAAAAAAAAAAKgCALACBMgCAtgCAPECZmZmZmZm5j%2F4AoQCkAMA%2BAITiAMAkgMEYmNkZQ%3D%3D\";window.tl_auction_response_644731={\"settings\":{\"advertiser_name\":\"TEST ADVERTISER NAME\",\"viewability\":{},\"impression_pixels\":[\"https:\\/\\/ad.doubleclick22222.net\\/ad\\/N4406.282142.ADARAMEDIAINC.\\/B8015471.114430570;sz=1x1;ord=142248259412135895294\"],\"type\":\"image\",\"additional_data\":{\"uid\":\"8947252401186410784\",\"brid\":\"86424\",\"bmid\":\"260\",\"aid\":\"21524535241656531860\",\"ts\":\"1560870964\",\"sid\":\"9\"},\"logo_image_url\":\"http:\\/\\/ep.yimg.com\\/ay\\/entirelypets\\/animal-planet-bat-dog-costume-large-21.jpg\",\"format_id\":1,\"render_options_bm\":0},\"assets\":[{\"image_id\":-1,\"image_url\":\"https:\\/\\/fublag.files.wordpress.com\\/2009\\/04\\/gen66337.png\",\"heading\":\"Renaissance Curacao Seat 3\",\"caption\":\"Discover the hidden gem of the Caribbean\",\"clickthrough_url\":\"http:\\/\\/ag.yieldoptimizer22222.com\\/ag\\/event?type=click&gi=vHH2MVL_SjnGKP62Cib00jK0s2o-vYF4fY9KN-ZMFBizsiWz6ErQ6BZVg_BuhEdSMwuOZFakknl2X_IMznDSZwult3EAcf-2t5bLa92B_a55wqXvj8LVNcUw5THdrYVDkzGglcgDIGNJBHMXOMtXE8627ymU1e3DANpWp1TQam6TMsQptW1GkKjvN34QhFK0&ref_url=123greetings.com\"}]};</script><script src=\"//ib.3lift.net/ttj?inv_code=makersalley_main\" data-auction-response-id=\"644731\"></script>",
          "adomain": [
            "marriott.com"
          ],
          "iurl": "https://fublag.files.wordpress.com/2009/04/gen66337.png",
          "crid": "260_9_2763332",
          "w": 300,
          "h": 250,
          "ext": {
            "prebid": {
              "type": "native"
            },
            "bidder": {
              "triplelift_pb": {
                "format": 1
              }
            }
          }
        }
      ],
      "seat": "triplelift"
    }
  ],
  "ext": {
    "responsetimemillis": {
      "triplelift": 121
    },
    "tmaxrequest": 500
  }
}"
```

### CODE REVIEW

**Comment by Kevin-P-Kerr** [comment]
not sure why the build is failing.  is there someway to retrigger?

running locally
```
[kkerr_local prebid-server]$ ./validate.sh
ok  	github.com/prebid/prebid-server	(cached)
ok  	github.com/prebid/prebid-server/adapters	(cached)
ok  	github.com/prebid/prebid-server/adapters/33across	(cached)
?   	github.com/prebid/prebid-server/adapters/adapterstest	[no test files]
ok  	github.com/prebid/prebid-server/adapters/adform	(cached)
ok  	github.com/prebid/prebid-server/adapters/adkernelAdn	(cached)
ok  	github.com/prebid/prebid-server/adapters/adtelligent	(cached)
ok  	github.com/prebid/prebid-server/adapters/appnexus	(cached)
ok  	github.com/prebid/prebid-server/adapters/audienceNetwork	(cached)
ok  	github.com/prebid/prebid-server/adapters/beachfront	(cached)
ok  	github.com/prebid/prebid-server/adapters/brightroll	(cached)
ok  	github.com/prebid/prebid-server/adapters/consumable	(cached)
ok  	github.com/prebid/prebid-server/adapters/conversant	(cached)
ok  	github.com/prebid/prebid-server/adapters/eplanning	(cached)
ok  	github.com/prebid/prebid-server/adapters/gamoshi	(cached)
ok  	github.com/prebid/prebid-server/adapters/grid	(cached)
ok  	github.com/prebid/prebid-server/adapters/gumgum	(cached)
ok  	github.com/prebid/prebid-server/adapters/improvedigital	(cached)
ok  	github.com/prebid/prebid-server/adapters/ix	(cached)
ok  	github.com/prebid/prebid-server/adapters/lifestreet	(cached)
ok  	github.com/prebid/prebid-server/adapters/mgid	(cached)
ok  	github.com/prebid/prebid-server/adapters/openx	(cached)
ok  	github.com/prebid/prebid-server/adapters/pubmatic	(cached)
ok  	github.com/prebid/prebid-server/adapters/pulsepoint	(cached)
ok  	github.com/prebid/prebid-server/adapters/rhythmone	(cached)
ok  	github.com/prebid/prebid-server/adapters/rtbhouse	(cached)
ok  	github.com/prebid/prebid-server/adapters/rubicon	(cached)
ok  	github.com/prebid/prebid-server/adapters/sharethrough	(cached)
ok  	github.com/prebid/prebid-server/adapters/somoaudience	(cached)
ok  	github.com/prebid/prebid-server/adapters/sonobi	(cached)
ok  	github.com/prebid/prebid-server/adapters/sovrn	(cached)
ok  	github.com/prebid/prebid-server/adapters/triplelift	(cached)
ok  	github.com/prebid/prebid-server/adapters/unruly	(cached)
ok  	github.com/prebid/prebid-server/adapters/visx	(cached)
ok  	github.com/prebid/prebid-server/adapters/vrtcal	(cached)
ok  	github.com/prebid/prebid-server/adapters/yieldmo	(cached)
?   	github.com/prebid/prebid-server/analytics	[no test files]
ok  	github.com/prebid/prebid-server/analytics/config	(cached)
ok  	github.com/prebid/prebid-server/analytics/filesystem	(cached)
?   	github.com/prebid/prebid-server/cache	[no test files]
ok  	github.com/prebid/prebid-server/cache/dummycache	(cached)
ok  	github.com/prebid/prebid-server/cache/filecache	(cached)
ok  	github.com/prebid/prebid-server/cache/postgrescache	(cached)
ok  	github.com/prebid/prebid-server/config	(cached)
ok  	github.com/prebid/prebid-server/currencies	(cached)
ok  	github.com/prebid/prebid-server/endpoints	(cached)
ok  	github.com/prebid/prebid-server/endpoints/info	(cached)
ok  	github.com/prebid/prebid-server/endpoints/openrtb2	(cached)
?   	github.com/prebid/prebid-server/errortypes	[no test files]
ok  	github.com/prebid/prebid-server/exchange	(cached)
ok  	github.com/prebid/prebid-server/gdpr	(cached)
ok  	github.com/prebid/prebid-server/macros	(cached)
ok  	github.com/prebid/prebid-server/openrtb_ext	(cached)
ok  	github.com/prebid/prebid-server/pbs	(cached)
ok  	github.com/prebid/prebid-server/pbsmetrics	(cached)
ok  	github.com/prebid/prebid-server/pbsmetrics/config	(cached)
ok  	github.com/prebid/prebid-server/pbsmetrics/prometheus	(cached)
?   	github.com/prebid/prebid-server/prebid	[no test files]
ok  	github.com/prebid/prebid-server/prebid_cache_client	(cached)
ok  	github.com/prebid/prebid-server/router	(cached)
ok  	github.com/prebid/prebid-server/server	(cached)
?   	github.com/prebid/prebid-server/ssl	[no test files]
ok  	github.com/prebid/prebid-server/stored_requests	(cached)
ok  	github.com/prebid/prebid-server/stored_requests/backends/db_fetcher	(cached)
ok  	github.com/prebid/prebid-server/stored_requests/backends/empty_fetcher	(cached)
ok  	github.com/prebid/prebid-server/stored_requests/backends/file_fetcher	(cached)
ok  	github.com/prebid/prebid-server/stored_requests/backends/http_fetcher	(cached)
?   	github.com/prebid/prebid-server/stored_requests/caches/cachestest	[no test files]
ok  	github.com/prebid/prebid-server/stored_requests/caches/memory	(cached)
?   	github.com/prebid/prebid-server/stored_requests/caches/nil_cache	[no test files]
ok  	github.com/prebid/prebid-server/stored_requests/config	(cached)
ok  	github.com/prebid/prebid-server/stored_requests/events	(cached)
ok  	github.com/prebid/prebid-server/stored_requests/events/api	(cached)
ok  	github.com/prebid/prebid-server/stored_requests/events/http	(cached)
ok  	github.com/prebid/prebid-server/stored_requests/events/postgres	(cached)
ok  	github.com/prebid/prebid-server/usersync	(cached)
ok  	github.com/prebid/prebid-server/usersync/usersyncers	(cached)
Running: go vet
```

**Comment by hhhjort** [comment]
Travis is failing for lack of code coverage. I see you have the `adapterstest.RunJSONBidderTest()` test defined, but you have not put any test cases in `triplelifttest` directory. Thus you don't have any tests running to cover your adapter code. You should define at least 1 JSON test (preferably multiple tests, covering possible variations and error conditions) using the expected requests/response you have pasted above.

**Comment by Kevin-P-Kerr** [comment]
> Kevin, your adapter lacks the directory `adapters/triplelift/triplelifttest/` where JSON test files should be located. These test cases get executed from `adapters/triplelift/triplelift_test.go` via the `RunJSONBidderTest` function but the current version of your adapter is not running
> any because this folder with it's subfolders still does not exist. An adapter must at leaset run tests from `adapters/triplelift/triplelifttest/exemplary/*.json`
> 
> ```
> </prebid/prebid-server/
> ▾ adapters/
>   ▾ triplelift/
>     ▾ triplelifttest/  ==> does not exist
>       ▸ exemplary/     ==> does not exist
>       ▸ supplemental/  ==> does not exist
>       ▸ params/race/   ==> does not exist
>     triplelift.go
>     triplelift_test.go
>     usersync.go
>     usersync_test.go
> ```

yep--currently working on this

**Comment by Kevin-P-Kerr** [comment]
@hhhjort  I've added some tests into the triplelifttest/exemplary directory.  can you take a look?

File: openrtb_ext/bidders.go [hunk 1]
```diff
@@ -135,7 +137,7 @@ func NewBidderParamsValidator(schemaDirectory string) (BidderParamValidator, err
 	for _, fileInfo := range fileInfos {
 		bidderName := strings.TrimSuffix(fileInfo.Name(), ".json")
 		if _, isValid := BidderMap[bidderName]; !isValid {
-			return nil, fmt.Errorf("File %s/%s does not match a valid BidderName.", schemaDirectory, fileInfo.Name())
+			return nil, fmt.Errorf("File %s/%s does not match a valid BidderName.", schemaDirectory, bidderName)
```
Comment (guscarreon) [inline openrtb_ext/bidders.go]: Please change line 140 to 
```
140             return nil, fmt.Errorf("File %s/%s does not match a valid BidderName.", schemaDirectory, fileInfo.Name())
```

File: adapters/triplelift/triplelift.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+func (a *TripleliftAdapter) MakeRequests(request *openrtb.BidRequest, extra *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	errs := make([]error, 0, len(request.Imp))
+	reqs := make([]*adapters.RequestData, 0, 1)
+	// copy the request, because we are going to mutate it
+	tlRequest := *request
+	// this will contain all the valid impressions
+	var validImps []openrtb.Imp
+	// pre-process the imps
+	for _, imp := range tlRequest.Imp {
```
[the remark this answers is not in the document]
Comment (Kevin-P-Kerr) [reply adapters/triplelift/triplelift.go]: Yeah, that makes sense.  Is there any error reporting we can expect once this is live?

File: adapters/triplelift/triplelift.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+	// get the triplelift extension
+	var ext adapters.ExtImpBidder
+	var tlext openrtb_ext.ExtImpTriplelift
+	if err := json.Unmarshal(imp.Ext, &ext); err != nil {
+		return err
+	}
+	if err := json.Unmarshal(ext.Bidder, &tlext); err != nil {
+		return err
+	}
+	imp.TagID = tlext.InvCode
```
[the remark this answers is not in the document]
Comment (Kevin-P-Kerr) [reply adapters/triplelift/triplelift.go]: yeah, that sounds like a good idea

[the remark this answers is not in the document]
Comment (Kevin-P-Kerr) [reply adapters/triplelift/triplelift.go]: Took a stab at this on line 41 above.  Let me know what you think

File: adapters/triplelift/triplelift.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Kevin-P-Kerr) [reply adapters/triplelift/triplelift.go]: See line 73 below--if there are no valid imps, we return `nil, []error` with an array of errors we encountered.

File: adapters/triplelift/triplelift.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (guscarreon) [reply adapters/triplelift/triplelift.go]: Check looks good

File: adapters/triplelift/triplelift.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (guscarreon) [reply adapters/triplelift/triplelift.go]: Awesome, thanks

File: adapters/triplelift/triplelift.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+	var tlext openrtb_ext.ExtImpTriplelift
+	if err := json.Unmarshal(imp.Ext, &ext); err != nil {
+		return err
+	}
+	if err := json.Unmarshal(ext.Bidder, &tlext); err != nil {
+		return err
+	}
+	if imp.Banner == nil && imp.Video == nil {
+		return fmt.Errorf("neither Banner nor Video object specified")
+	}
```
Comment (guscarreon) [inline adapters/triplelift/triplelift.go]: If we accept banners, are the width and height properties relevant? Other adapters check for `nil` on the `imp.Banner.W`, `imp.Banner.H`, and `imp.Banner.Format`. Would that be useful in our context?
```
 34 func processImp(imp *openrtb.Imp) error {
 35     // get the triplelift extension
 36     var ext adapters.ExtImpBidder
 37     var tlext openrtb_ext.ExtImpTriplelift
 38     if err := json.Unmarshal(imp.Ext, &ext); err != nil {
 39         return err
 40     }
 41     if err := json.Unmarshal(ext.Bidder, &tlext); err != nil {
 42         return err
 43     }
 44     if imp.Banner == nil && imp.Video == nil {
 45         return fmt.Errorf("neither Banner nor Video object specified")
 46     }
    +   if imp.Banner {
    +   	if imp.Banner.W == nil && imp.Banner.H == nil && len(imp.Banner.Format) > 0 {
    +              firstFormat := bannerCopy.Format[0]
    +              bannerCopy.W = &(firstFormat.W)
    +              bannerCopy.H = &(firstFormat.H)
    +          }
    +    }
 47     imp.TagID = tlext.InvCode
 48     // floor is optional
 49     if tlext.Floor == nil {
```
Maybe width and height are not relevant at this point, just asking.

Comment (Kevin-P-Kerr) [reply adapters/triplelift/triplelift.go]: Good point, but this is something we'd prefer to handle downstream from this application.

**Comment by Kevin-P-Kerr** [comment]
> github.com/prebid/prebid-server/adapters/triplelift/triplelift.go:58:   MakeRequests            73.9%

will do

**Comment by Kevin-P-Kerr** [comment]
Hey, @guscarreon , it seems that most of the lines not covered by the unit tests cover error conditions, like

```
 if response.StatusCode != http.StatusOK {
        return nil, []error{fmt.Errorf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode)}
    }
```
Even though it's just a few lines, since these functions are all so short, that they aren't covered by the test procedures reduces code coverage pretty significantly.   
But if I were to add a broken a request to the test, the test would fail.  
Do you have any insight into how to test error conditions?

**Comment by guscarreon** [comment]
> Hey, @guscarreon , it seems that most of the lines not covered by the unit tests cover error conditions, like
> 
> ```
>  if response.StatusCode != http.StatusOK {
>         return nil, []error{fmt.Errorf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode)}
>     }
> ```
> 
> Even though it's just a few lines, since these functions are all so short, that they aren't covered by the test procedures reduces code coverage pretty significantly.
> But if I were to add a broken a request to the test, the test would fail.
> Do you have any insight into how to test error conditions?

Hi @Kevin-P-Kerr. I believe the [EqualError()](https://godoc.org/github.com/stretchr/testify/assert#EqualError) function is useful for this purpose. Some test cases throughout the code base use it extensively:
```
./config/config_test.go:        assert.EqualError(t, errs[0], message)
./openrtb_ext/device_test.go:   assert.EqualError(t, json.Unmarshal([]byte(`{"prebid":{"interstitial":{"minwidthperc":85,"minheightperc":"75"}}}`), &s), "request.device.ext.prebid.interstitial.minheightperc must be a number between 0 and 100")
./openrtb_ext/site_test.go:     assert.EqualError(t, json.Unmarshal([]byte(`{"amp":-1}`), &s), "request.site.ext.amp must be either 1, 0, or undefined")
./adapters/info_test.go:        assert.EqualError(t, errs[0], "this bidder does not support app requests")
```

**Comment by Kevin-P-Kerr** [comment]
@guscarreon   cumulative test coverage is now 95 percent.

```
[kkerr_local triplelift]$ go test -coverprofile=c.out
PASS
coverage: 95.5% of statements
ok  	github.com/prebid/prebid-server/adapters/triplelift	0.017s
[kkerr_local triplelift]$ go tool cover -func=c.out
github.com/prebid/prebid-server/adapters/triplelift/triplelift.go:26:	getBidType		100.0%
github.com/prebid/prebid-server/adapters/triplelift/triplelift.go:34:	processImp		100.0%
github.com/prebid/prebid-server/adapters/triplelift/triplelift.go:58:	MakeRequests		91.3%
github.com/prebid/prebid-server/adapters/triplelift/triplelift.go:96:	getBidCount		100.0%
github.com/prebid/prebid-server/adapters/triplelift/triplelift.go:104:	MakeBids		95.2%
github.com/prebid/prebid-server/adapters/triplelift/triplelift.go:144:	NewTripleliftBidder	100.0%
github.com/prebid/prebid-server/adapters/triplelift/usersync.go:10:	NewTripleliftSyncer	100.0%
total:									(statements)		95.5%
```

**Comment by guscarreon** [comment]
> @guscarreon cumulative test coverage is now 95 percent.
> 
> ```
> [kkerr_local triplelift]$ go test -coverprofile=c.out
> PASS
> coverage: 95.5% of statements
> ok  	github.com/prebid/prebid-server/adapters/triplelift	0.017s
> [kkerr_local triplelift]$ go tool cover -func=c.out
> github.com/prebid/prebid-server/adapters/triplelift/triplelift.go:26:	getBidType		100.0%
> github.com/prebid/prebid-server/adapters/triplelift/triplelift.go:34:	processImp		100.0%
> github.com/prebid/prebid-server/adapters/triplelift/triplelift.go:58:	MakeRequests		91.3%
> github.com/prebid/prebid-server/adapters/triplelift/triplelift.go:96:	getBidCount		100.0%
> github.com/prebid/prebid-server/adapters/triplelift/triplelift.go:104:	MakeBids		95.2%
> github.com/prebid/prebid-server/adapters/triplelift/triplelift.go:144:	NewTripleliftBidder	100.0%
> github.com/prebid/prebid-server/adapters/triplelift/usersync.go:10:	NewTripleliftSyncer	100.0%
> total:									(statements)		95.5%
> ```

Kevin, some Travis errors seem to have popped up with this last update, can we correct them?

**Comment by Kevin-P-Kerr** [comment]
> > @guscarreon cumulative test coverage is now 95 percent.
> > ```
> > [kkerr_local triplelift]$ go test -coverprofile=c.out
> > PASS
> > coverage: 95.5% of statements
> > ok  	github.com/prebid/prebid-server/adapters/triplelift	0.017s
> > [kkerr_local triplelift]$ go tool cover -func=c.out
> > github.com/prebid/prebid-server/adapters/triplelift/triplelift.go:26:	getBidType		100.0%
> > github.com/prebid/prebid-server/adapters/triplelift/triplelift.go:34:	processImp		100.0%
> > github.com/prebid/prebid-server/adapters/triplelift/triplelift.go:58:	MakeRequests		91.3%
> > github.com/prebid/prebid-server/adapters/triplelift/triplelift.go:96:	getBidCount		100.0%
> > github.com/prebid/prebid-server/adapters/triplelift/triplelift.go:104:	MakeBids		95.2%
> > github.com/prebid/prebid-server/adapters/triplelift/triplelift.go:144:	NewTripleliftBidder	100.0%
> > github.com/prebid/prebid-server/adapters/triplelift/usersync.go:10:	NewTripleliftSyncer	100.0%
> > total:									(statements)		95.5%
> > ```
> 
> Kevin, some Travis errors seem to have popped up with this last update, can we correct them?

hey @guscarreon  I think github might have just froze up or something.

```
 Cloning into '/home/travis/gopath/pkg/dep/sources/https---github.com-stretchr-objx'...
485fatal: unable to access 'https://github.com/stretchr/objx/': Failed to connect to github.com port 443: Connection timed out
486: command failed: [git clone --recursive -v --progress https://github.com/stretchr/objx /home/travis/gopath/pkg/dep/sources/https---github.com-stretchr-objx]: exit status 128
487The command "$GOPATH/bin/dep ensure" failed and exited with 1 during .
```

seems like it couldn't access this strechr-objx dependency.  Anyway we can retrigger?

**Comment by Kevin-P-Kerr** [comment]
> @Kevin-P-Kerr seems like Github was having problems because for some reason the Travis build seems to have built correctly this time around. Thanks for your patience, your adapter looks awesome.

Thanks!  What are the next steps in this process?

File: adapters/triplelift/triplelift.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+	imp.TagID = tlext.InvCode
+	// floor is optional
+	if tlext.Floor == nil {
```
Comment (hhhjort) [inline adapters/triplelift/triplelift.go]: Could just have `if tlext.Floor != nil { imp.BidFloor = *tlext.Floor }` as you return `nil` in either case

Comment (Kevin-P-Kerr) [reply adapters/triplelift/triplelift.go]: ah good point, that's a bit more parsimonious.  but it seems this is already merged. if I have to patch later I'll take this idea and apply it though.

**Comment by hhhjort** [comment]
@Kevin-P-Kerr : Just noticed that the user sync URL makes no reference to the PBS URL ... there is no way for your usersync endpoint to send the user ID back to the particular PBS installation.

**Comment by Kevin-P-Kerr** [comment]
@hhhjort  I'm not totally following.  Can you elaborate?  What steps should I take to rectify this?

**Comment by hhhjort** [comment]
Look at https://github.com/prebid/prebid-server/blob/master/docs/developers/cookie-syncs.md

Basically, your cookie sync endpoint needs to redirect back to the supplied URL and include the user's Triplelift ID so that PBS can populate `user.id` in the request with your ID. This will allow you to do user targeting. You cannot rely on your own cookie, as the bid request will come from the PBS machine, and it does not have access to the user's triplelift cookie.

**Comment by Kevin-P-Kerr** [comment]
> Look at https://github.com/prebid/prebid-server/blob/master/docs/developers/cookie-syncs.md
> 
> Basically, your cookie sync endpoint needs to redirect back to the supplied URL and include the user's Triplelift ID so that PBS can populate `user.id` in the request with your ID. This will allow you to do user targeting. You cannot rely on your own cookie, as the bid request will come from the PBS machine, and it does not have access to the user's triplelift cookie.

Ok, thanks for the clarification.  We'll schedule some time to get a fix in.

**Pull request merged**